### ISSUE TITLE
train_imitation script suddenly Killed

### ISSUE THREAD

**Issue by hhroberthdaniel** [HumanCompatibleAI/imitation#401]
Hi,

I have installed the latest commit and everything seemed to work fine until I tried training for longer time period.
After training for ~30 min the train_imitation.py used with dagger suddenly stops. The output is just this:

`INFO - root - Loading demonstrations`
`INFO - root - Loaded 10530 total`
`Killed`

Unfortunately, I am runnning this with a custom environment so it is harder to reproduce.
Do you have any idea why this might have happened?
I checked and the total_timesteps has not been reached. I also bumped it to 1e10 and the process has been stopped after around the same amount of time.

Do you have any suggestions where I can look into?

**Comment by AdamGleave** [comment]
Killed indicates the process received a signal to terminate from the operating system.

Running out of memory is a common cause of this, worth checking `dmesg` for OOM errors. If there's a memory leak in DAgger then feel free to open a bug for that, but I think it's normal memory consumption would increase over time with that algorithm, as it interactively collects a growing dataset.

**Comment by hhroberthdaniel** [comment]
You are right. Is this behaviour normal? I find it a bit odd to have an algorithm that would eventually throw an OOM error if ran for enough time. Perhaps a simple variable to control how many of the last demos to load? Or even better, to save each transition on the disk and use a pytorch dataloader (with multiple workers) to load each batch?
I would not mind implementing it myself

**Comment by AdamGleave** [comment]
I agree it'd be better to limit the amount of RAM used over time, either by a max dataset size and dropping data beyond that point (e.g. the oldest, or just drop points at random), or by saving to disk and streaming data as needed.

In particular, I think we could run out of RAM even with data that a human could realistically produce. For example, one hour of a human driving in [Carla](https://carla.readthedocs.io/) would produce 144GB of data (800x600x3 bytes per frame). This is already bigger than many machines can handle.

I don't think this is causing us issues with our environments, so not a priority for us to fix, but I'd be happy to review & merge a PR addressing this.

**Comment by AdamGleave** [comment]
Closing due to inactivity, but PR on this issue is still welcome if this is causing you (or others) problems.

**Comment by benediktkreis** [comment]
I'm running into the same problem after just 4160 timesteps (~7 rounds) on a machine with 128 GB RAM. What would be an adequate window size for dropping old demonstrations?
@hhroberthdaniel: Did you find a solution or did you implement a custom strategy to avoid the memory issue?

**Comment by paulo101977** [comment]
> I'm running into the same problem after just 4160 timesteps (~7 rounds) on a machine with 128 GB RAM. What would be an adequate window size for dropping old demonstrations? [@hhroberthdaniel](https://github.com/hhroberthdaniel): Did you find a solution or did you implement a custom strategy to avoid the memory issue?

I'm having the same problem as the post. 64 GB of RAM and after 7 Rounds. I don't know if the problem is actually memory, as I had 32 GB free.

**Comment by paulo101977** [comment]
> You are right. Is this behaviour normal? I find it a bit odd to have an algorithm that would eventually throw an OOM error if ran for enough time. Perhaps a simple variable to control how many of the last demos to load? Or even better, to save each transition on the disk and use a pytorch dataloader (with multiple workers) to load each batch?
> I would not mind implementing it myself

I'm getting the same error here, always in the same place, and I have enough memory. I think the problem isn't memory. I'll try to debug it.
